OpenRA.Game/Map/Map.cs

```diff
@@ -1068,20 +1071,44 @@ public MPos ChooseClosestEdgeCell(MPos uv)
 			return edge.V == Bounds.Bottom ? unProjected.MaxBy(x => x.V) : unProjected.MinBy(x => x.V);
 		}
 
-		public CPos ChooseRandomEdgeCell(MersenneTwister rand)
+		public CPos ChooseClosestMatchingEdgeCell(CPos cell, Func<CPos, bool> match)
 		{
-			List<MPos> cells;
-			do
+			return AllEdgeCells.OrderBy(c => (cell - c).Abs().Length).FirstOrDefault(c => match(c));
```

Comment:
The `Abs()` isn't needed because the length calculation doesn't care about the sign of the components.